kernel/thread/thread.c

```diff
@@ -838,6 +843,10 @@ int thd_join(kthread_t *thd, void **value_ptr) {
     if(thd == NULL)
         return -1;
 
+    /* If you wait for yourself, you'll never leave */
+    if(thd == thd_current)
+        return -4;
```

Comment:
This allows the graceful cleanup of `thd_join`ing threads that the system is using without having to worry if that thread itself was what requested the shutdown. I think generally as well, it's a good idea to avoid possible deadlocks.

At some point we likely need to refactor the exit paths in order for them to be assured to run in one of the system threads.